core/consensus/grandpa/impl/voting_round_impl.cpp

```diff
@@ -545,322 +747,560 @@ namespace kagome::consensus::grandpa {
-              // ancestry of p-G we vote for the best chain containing it.
-              if (auto ancestry =
-                      env_->getAncestry(last_round_estimate.block_hash,
-                                        last_prevote_g.block_hash);
-                  ancestry) {
-                auto to_sub = primary.block_number + 1;
-
-                primitives::BlockNumber offset = 0;
-                if (last_prevote_g.block_number >= to_sub) {
-                  offset = last_prevote_g.block_number - to_sub;
-                }
+  bool VotingRoundImpl::updateCompletability() {
+    // figuring out whether a block can still be committed for is
+    // not straightforward because we have to account for all possible future
+    // equivocations and thus cannot discount weight from validators who
+    // have already voted.
+    const auto tolerated_equivocations = voter_set_->totalWeight() - threshold_;
 
-                if (ancestry.value().at(offset) == primary.block_hash) {
-                  return primary;
-                }
-              }
-              return last_round_estimate;
-            })
-            .value_or_eval(
-                [last_round_estimate] { return last_round_estimate; });
+    // find how many more equivocations we could still get.
+    //
+    // it is only important to consider the voters whose votes
+    // we have already seen, because we are assuming any votes we
+    // haven't seen will target this block.
 
-    auto rbest_chain = env_->bestChainContaining(find_descendent_of.block_hash);
+    // get total weight of all equivocators
+    const auto current_equivocations = std::accumulate(
+        precommit_equivocators_.begin(),
+        precommit_equivocators_.end(),
+        0ul,
+        [this, index = 0ul](size_t sum, auto isEquivocator) mutable {
+          if (not isEquivocator) {
+            ++index;
+            return sum;
+          }
+          return sum + voter_set_->voterWeight(index++).value();
+        });
 
-    if (not rbest_chain) {
-      logger_->error(
-          "Could not cast prevote: previously known block {} has disappeared",
-          find_descendent_of.block_hash.toHex());
-      return outcome::failure(boost::system::error_code());
+    const auto additional_equivocations =
+        tolerated_equivocations - current_equivocations;
+
+    auto possible_to_precommit = [&](const VoteWeight &weight) {
+      // total precommits for this block, including equivocations.
+      auto precommited_for =
+          weight
+              .totalWeight(
+                  prevote_equivocators_, precommit_equivocators_, voter_set_)
+              .precommit;
+
+      // equivocations we could still get are out of those who
+      // have already voted, but not on this block.
+      auto possible_equivocations =
+          std::min<size_t>(precommits_->getTotalWeight() - precommited_for,
+                           additional_equivocations);
+
```

Comment:
It seems ```precommits_->getTotalWeight()``` can be less than ```precommited_for```.

Reply:
Diffenetely not. _getTotalWeight_ means summary weight of all one stage's vote and no one block can have greater weight.